packages/peregrine/lib/talons/RootComponents/Product/useProduct.js

```diff
@@ -20,70 +21,52 @@ import { useApolloClient, useQuery } from '@apollo/react-hooks';
+    const { mapProduct, typePolicies, queries, urlKey } = props;
 
     const apolloClient = useApolloClient();
+    apolloClient.cache.policies.addTypePolicies(typePolicies);
 
-    /*
-     * Look up the product in cache first.
-     *
-     * We may have it from a previous query, but we have to manually tell Apollo cache where to find it.
-     * A single product query (as described by https://github.com/magento/graphql-ce/issues/86)
-     * will alleviate the need to do this manual work.
-     *
-     * If the object with the specified `id` is not in the cache, we get `null`.
-     * If the object is in the cache but doesn't have all the fields we need, an error is thrown.
-     *
-     * @see https://www.apollographql.com/docs/react/caching/cache-interaction/#readfragment.
-     */
-    const productFromCache = useMemo(() => {
-        try {
-            return apolloClient.readFragment({
-                id: `${cachePrefix}:${urlKey}`,
-                fragment
-            });
-        } catch (e) {
-            // The product is in the cache but it is missing some fields the fragment needs.
-            return null;
+    const {
+        error: cacheError,
+        loading: cacheLoading,
+        data: cacheData
+    } = useQuery(queries.getProductFromCache, {
+        variables: {
+            urlKey
         }
-    }, [apolloClient, cachePrefix, fragment, urlKey]);
+    });
 
-    /*
-     * Fetch the product from the network.
-     *
-     * Note that this always fires, regardless of whether we have a cached product or not:
-     * If we do not have a cached product, we need to go fetch it from the network.
-     * If we do have a cached product, we want to ensure its cache entry doesn't get stale.
-     * In both cases, Apollo will update the cache with the latest data when this returns.
-     */
-    const { loading, error, data } = useQuery(query, {
-        // Once we're able to remove the manual cache lookup,
-        // this fetch policy can change to 'cache-and-network'.
-        fetchPolicy: 'network-only',
+    const {
+        error: networkError,
+        loading: networkLoading,
+        data: networkData
+    } = useQuery(queries.getProductFromNetwork, {
+        fetchPolicy: 'cache-and-network',
+        nextFetchPolicy: 'cache-first',
         variables: {
             urlKey
         }
     });
 
     const product = useMemo(() => {
-        if (productFromCache) {
-            return mapProduct(productFromCache);
+        if (cacheData && cacheData.product) {
+            return mapProduct(cacheData.product);
         }
 
         // If a product is out of stock _and_ the backend specifies not to
         // display OOS items, the items array will be empty.
-        if (data && data.productDetail.items[0]) {
-            const productFromNetwork = data.productDetail.items[0];
+        if (networkData && networkData.products.items[0]) {
+            const productFromNetwork = networkData.products.items[0];
```

Comment:
I saw elsewhere that Apollo is more assertive about avoiding mutation of cache data. Do we need to emphasize that `mapProduct` must not mutate the data it receives, since we're passing it the original data? Or is this unrelated?

Reply:
Totally valid point - we happen to map into a new object, not over the existing, but it would be good to add that note to the function doc. I'll make sure to update it.